File: ruby/raindrops/raindrops.rb
require 'Prime'

# Raindrops
class Raindrops
  class << self
    WORDS = {
      3 => 'Pling',
      5 => 'Plang',
      7 => 'Plong'
    }.freeze
    def convert(num)
      return num.to_s unless valid?(factors(num))
      translate(factors(num))
    end

    def factors(num)
      Prime.prime_division(num).flatten.uniq.sort
    end

    def valid?(factors)
      factors.include?(3) || factors.include?(5) || factors.include?(7)
    end

    def translate(factors)
      factors.map { |factor| translate_factor(factor) }.join
    end

    def translate_factor(factor)
      WORDS.include?(factor) ? WORDS[factor] : ''
    end
  end
end

File: java/raindrops/src/main/java/Raindrops.java
import java.util.ArrayList;
import java.util.List;

public class Raindrops {

  public static String convert(int input) {
    List<Integer> factors = getForNumber(input);

    StringBuilder sb = new StringBuilder();
    if (factors.contains(3)) {
      sb.append("Pling");
    }
    if (factors.contains(5)) {
      sb.append("Plang");
    }
    if (factors.contains(7)) {
      sb.append("Plong");
    }
    if (sb.toString().isEmpty()) {
      sb.append(Integer.toString(input));
    }

    return sb.toString();
  }

  public static List<Integer> getForNumber(int input) {
    List<Integer> factors = new ArrayList<Integer>();
    int num = input;

    for (int i = 2; i <= num; i++) {
      if (num % i == 0) {
        factors.add(i);
        num = num / i;
        i = i - 1;
      }
    }

    return factors;
  }
}
